### PR TITLE
Add Stream IO

### PR DESCRIPTION
**Pull request by ianshmean**
I got as far as I could with load methods for streams, and got stuck so I'm leaving my progress up here.

I've been following: http://www.piko3d.net/tutorials/libpng-tutorial-loading-png-files-from-streams/

Anyone know what to do about
https://github.com/JuliaIO/PNGFiles.jl/blob/stream_io/src/io.jl#L54

I added the appropriate methods to ImageIO in anticipation of this working: https://github.com/JuliaIO/ImageIO.jl/pull/3

cc. @Drvi @SimonDanisch

### CODE REVIEW

File: src/io.jl [hunk 1]
```diff
@@ -22,6 +22,39 @@ function load(fpath::String; gamma::Union{Nothing,Float64}=nothing, expand_palet
...
+        return IndirectArray(buffer, OffsetArray(palette, -1))
+    end
+end
+function _readcallback(pngPtr::png_structp, data::png_bytep, length::png_size_t)
+    # Here we get our IO pointer back from the read struct.
+    # This is the parameter we passed to the png_set_read_fn() function.
+    # Our std::istream pointer.
+    a = png_get_io_ptr(pngPtr)
+    # Cast the pointer to std::istream* and read 'length' bytes into 'data'
+    ((std::istream*)a)->read(data, length); #TODO: THIS IS WRONG. Not sure what to do here
```
Comment (ianshmean) [inline src/io.jl:54]: Not sure what to do here

**Comment by Drvi** [comment]
This seems to work, but I was more playing segfault whack-a-mole rather than knowing what I'm supposed to do. Someone needs to check the use of `ccall(:ios_readll, ...)` bit, as I can see that in [Base](https://github.com/JuliaLang/julia/blob/97a62584750c4f8d23ad57305a65721074f7038e/base/iostream.jl#L447), it is used within `@_lock_ios` block.

**Comment by ianshmean** [comment]
Nice work! Thanks for figuring that out!
I do think the lock is a good idea, so I've added it (via the pre 1.5 approach of `lock(io.lock)` rather than the new macro).

I think we should add the corresponding `save` methods for IO's, but I've found less guidance on that online. I'll keep looking

**Comment by ianshmean** [comment]
Success! Load and save methods now available for IO streams.

@Drvi if you're happy with this, we can merge tomorrow and release, along with the support in ImageIO https://github.com/JuliaIO/ImageIO.jl/pull/3

**Comment by Drvi** [comment]
Thanks for looking into the `save(s::IO, ...)` and locking business!

I noticed that I could silently write into a read-only stream and that it segfaults when reading an already processed stream, so I added a couple of checks.

Other than that, I think this is good to go!

**Pull request merged**